xds/src/main/java/io/grpc/xds/WeightedTargetLoadBalancer.java

```diff
@@ -0,0 +1,216 @@
+import io.grpc.xds.WeightedTargetLoadBalancerProvider.WeightedTargetConfig;
+import io.grpc.xds.XdsLogger.XdsLogLevel;
+import io.grpc.xds.XdsSubchannelPickers.ErrorPicker;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/** Load balancer for weighted_target policy. */
+final class WeightedTargetLoadBalancer extends LoadBalancer {
+
+  private final XdsLogger logger;
+  private final LoadBalancerRegistry lbRegistry;
+  private final Map<String, GracefulSwitchLoadBalancer> childBalancers = new HashMap<>();
+  private final Map<String, ChildHelper> childHelpers = new HashMap<>();
+  private final Helper helper;
+  private final WeightedPickerFactory weightedPickerFactory;
+
+  private Map<String, WeightedChildLbConfig> targets = ImmutableMap.of();
+
+  /**
+   * Constructs a WeightedTargetLoadBalancer with the given weighted list of child balancer configs.
+   * The list must not be empty and must not contain duplicate lb configs.
+   */
+  WeightedTargetLoadBalancer(Helper helper, LoadBalancerRegistry lbRegistry) {
+    this(
+        checkNotNull(helper, "helper"),
+        checkNotNull(lbRegistry, "lbRegistry"),
+        WeightedPickerFactory.RANDOM_PICKER_FACTORY);
+  }
+
+  @VisibleForTesting
+  WeightedTargetLoadBalancer(
+      Helper helper, LoadBalancerRegistry lbRegistry, WeightedPickerFactory weightedPickerFactory) {
+    this.helper = helper;
+    this.lbRegistry = lbRegistry;
+    this.weightedPickerFactory = weightedPickerFactory;
+    logger = XdsLogger.withLogId(InternalLogId.allocate("cds-lb", helper.getAuthority()));
+    logger.log(XdsLogLevel.INFO, "Created");
+  }
+
+  @Override
+  public void handleResolvedAddresses(ResolvedAddresses resolvedAddresses) {
+    logger.log(XdsLogLevel.DEBUG, "Received resolution result: {0}", resolvedAddresses);
+    Object lbConfig = resolvedAddresses.getLoadBalancingPolicyConfig();
+    checkNotNull(lbConfig, "missing weighted_target lb config");
+
+    WeightedTargetConfig weightedTargetConfig = (WeightedTargetConfig) lbConfig;
+    Map<String, WeightedChildLbConfig> newTargets = weightedTargetConfig.targets;
+
+    for (String targetName : newTargets.keySet()) {
+      WeightedChildLbConfig weightedChildLbConfig = newTargets.get(targetName);
+      if (!targets.containsKey(targetName)) {
+        // Create child balancers for new names.
+        ChildHelper childHelper = new ChildHelper();
+        GracefulSwitchLoadBalancer childBalancer = new GracefulSwitchLoadBalancer(childHelper);
+        childBalancer.switchTo(lbRegistry.getProvider(weightedChildLbConfig.policyName));
+        childHelpers.put(targetName, childHelper);
+        childBalancers.put(targetName, childBalancer);
+      } else if (!weightedChildLbConfig.policyName.equals(targets.get(targetName).policyName)) {
+        // Policy name change for the same target is not a typical usecase, but we should support
```

Comment:
nit: commenting is redundant. Same for above and below. Code is already clearly saying what it is doing.